### DISCUSSION TITLE
Can progressbars be rendered in parallel?

### DISCUSSION THREAD

**Discussion by sheldonhull** [pterm/pterm#254] [Q&A]
I wanted to have 3 goroutines run at the same time showing progress.
I am assuming this is currently not supported, but wanted to ask here before opening a feature request. 

Example code modified from [uiprogress](https://github.com/gosuri/uiprogress) for pterm. It doesn't work, but maybe I'm missing something?

```go
package main

import (
	"sync"
	"time"

	"github.com/pterm/pterm"
)

func main() {
	waitTime := time.Millisecond * 100

	// start the progress bars in go routines
	var wg sync.WaitGroup
	bar1, _ := pterm.DefaultProgressbar.WithTotal(100).WithTitle("Downloading more ram").Start()

	// bar1 := uiprogress.AddBar(20).AppendCompleted().PrependElapsed()
	wg.Add(1)
	go func() {
		defer wg.Done()
		for i := 0; i < 100; i++ {
			bar1.Increment() // Increment the progressbar by one. Use Add(x int) to increment by a custom amount.
			time.Sleep(waitTime)
		}
	}()

	bar2, _ := pterm.DefaultProgressbar.WithTotal(100).WithTitle("Contemplating things").Start()
	wg.Add(1)
	go func() {
		defer wg.Done()
		for i := 0; i < 100; i++ {
			bar2.Increment() // Increment the progressbar by one. Use Add(x int) to increment by a custom amount.
			time.Sleep(waitTime)
		}
	}()

	time.Sleep(time.Second)
	bar3, _ := pterm.DefaultProgressbar.WithTotal(100).WithTitle("Reticulating Splines").Start()
	wg.Add(1)
	go func() {
		defer wg.Done()
		for i := 0; i < 100; i++ {
			bar3.Increment() // Increment the progressbar by one. Use Add(x int) to increment by a custom amount.
			time.Sleep(waitTime)
		}
	}()

	// wait for all the go routines to finish
	wg.Wait()
}

```

**Comment by MarvinJWendt** [comment]
Yes you're right. This is currently not supported, as we use the ANSI code to rewrite the last line in the terminal (`\r`). But with our "new" `AreaPrinter` we can reprint multiple lines, so this is now possible. You can make a FR 😄 
It would have the same logic as #215

**Reply by sheldonhull** [reply]
@MarvinJWendt  awesome! I'll take a look at the example and if this works I'll mark as solved.

Sorry, what is an FR? I think you meant PR?
I'll take a look, not sure yet if my Go chops are quite up to that level, I'll have to look a bit more. 
Maybe if you add a few pointers on what might need to change on the issue mentioned, it would help a newer contributor not quite as familiar with the Pterm project?

**Reply by MarvinJWendt** [reply]
Hi @sheldonhull, a FR is a "Feature Request". I meant, you can write an issue :)
Everyone is free to create a PR then, but I'll most likely do it myself, as it's kinda complicated and requires some more refactoring ^^

**Comment by kinthaiofficial** [comment]
Parallel progressbars in PTerm is possible, though it requires some care with concurrent rendering. Here's a working pattern:

```go
package main

import (
    "github.com/pterm/pterm"
    "sync"
    "time"
)

func main() {
    // Create a multi-printer for parallel progress bars
    multi := pterm.DefaultMultiPrinter

    // Create progress bars within the multi-printer
    pb1, _ := pterm.DefaultProgressbar.WithTotal(100).
        WithWriter(multi.NewWriter()).Start("Task 1")
    pb2, _ := pterm.DefaultProgressbar.WithTotal(100).
        WithWriter(multi.NewWriter()).Start("Task 2")
    pb3, _ := pterm.DefaultProgressbar.WithTotal(100).
        WithWriter(multi.NewWriter()).Start("Task 3")

    multi.Start()
    defer multi.Stop()

    var wg sync.WaitGroup
    
    // Run tasks concurrently
    for i, pb := range []*pterm.ProgressbarPrinter{pb1, pb2, pb3} {
        wg.Add(1)
        go func(pb *pterm.ProgressbarPrinter, delay int) {
            defer wg.Done()
            for j := 0; j < 100; j++ {
                pb.Increment()
                time.Sleep(time.Duration(delay) * time.Millisecond)
            }
        }(pb, (i+1)*10)
    }
    
    wg.Wait()
}
```

**Key pattern: `pterm.DefaultMultiPrinter`**
This is the component that handles concurrent rendering — it serializes terminal writes so bars don't corrupt each other. Each progress bar gets its own `Writer` from the multi-printer.

**Performance note:**
With many parallel bars (10+), the re-render frequency matters. PTerm renders on each `Increment()` call — if your goroutines update very frequently, consider batching updates (increment every N steps rather than every step).

Is this for an agent orchestration monitoring UI? That's a great use case for parallel progress bars.
